cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/device_coo_matrix.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/linalg/map_reduce.cuh>
+#include <raft/matrix/init.cuh>
+#include <raft/sparse/convert/coo.cuh>
+#include <raft/sparse/neighbors/cross_component_nn.cuh>
+#include <raft/sparse/op/sort.cuh>
+
+#include <thrust/reduce.h>
+
+namespace raft::sparse::matrix::detail {
+
+/**
+ * This function creates a representation of input data (rows) that identifies when
+ * value changes in the input data. This function assumes data is sorted.
+ *
+ * @param[in] rows
+ *   The input data
+ * @param[in] nnz
+ *   The size of the input data.
+ * @param[in] counts
+ *   The resulting representation of the index value changes of the input. Should be
+ *   the same size as the input (nnz)
+ */
+__global__ void _scan(int* rows, int nnz, int* counts)
+{
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index >= nnz) { return; }
+  if (index == 0) {
+    counts[index] = 1;
+    return;
+  }
+  if (index < nnz) {
+    int curr_id = rows[index];
+    int old_id  = rows[index - 1];
+    if (curr_id != old_id) {
+      counts[index] = 1;
+    } else {
+      counts[index] = 0;
+    }
+  }
+}
+
+/**
+ * This function counts the occurrences of the input array. Uses modulo logic as a
+ * rudimentary hash (should be changed with better hash function).
+ *
+ * @param[in] cols
+ *   The input data
+ * @param[in] nnz
+ *   The size of the input data.
+ * @param[in] counts
+ *   The resulting representation of the index value changes of the input. Should be
+ *   the same size as the input (nnz)
+ * @param[in] feats
+ *   The array that will house the occurrence counts
+ * @param[in] vocabSize
+ *   The size of the occurrence counts array (feats).
+ */
+__global__ void _fit_compute_occurs(int* cols, int nnz, int* counts, int* feats, int vocabSize)
+{
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  if ((index < nnz) && (counts[index] == 1)) {
+    int targetVal = cols[index];
+    int vocab     = targetVal % vocabSize;
```

Comment:
This line is the equivalent to hash, when we have the right function for the hash we can replace here. This allows number of features to come in. Even if the actual class was built with only a subset of features allowed.